File: library/src/main/java/me/shkschneider/skeleton/helper/ApplicationHelper.java
package me.shkschneider.skeleton.helper;

import android.annotation.SuppressLint;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.Size;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.shkschneider.skeleton.java.SkHide;
import me.shkschneider.skeleton.ui.BitmapHelper;
import me.shkschneider.skeleton.SkeletonApplication;

public class ApplicationHelper {

    public static final int DEFAULT_ICON = android.R.drawable.sym_def_app_icon;
    public static final String INSTALLER = "com.android.vending";

    protected ApplicationHelper() {
        // Empty
    }

    public static boolean debug() {
        // <https://stackoverflow.com/a/25517680/603270>
        return SkeletonApplication.DEBUG;
    }

    public static Resources resources() {
        return ContextHelper.applicationContext().getResources();
    }

    @Deprecated //Avoid
    public static AssetManager assetManager() {
        return AssetsHelper.assetManager();
    }

    @Deprecated // Discouraged
    public static String[] files() {
        return ContextHelper.applicationContext().fileList();
    }

    public static String packageName() {
        return ContextHelper.applicationContext().getPackageName();
    }

    public static PackageManager packageManager() {
        return ContextHelper.applicationContext().getPackageManager();
    }

    @Nullable
    public static String name() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }

            final CharSequence label = applicationInfo.loadLabel(packageManager);
            if (label == null) {
                LogHelper.warning("Label was NULL");
                return null;
            }

            return label.toString();
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    // <http://semver.org>
    @Size(3)
    @Nullable
    public static Integer[] semanticVersions() {
        final String versionName = versionName();
        if (versionName == null) {
            return null;
        }
        if (! versionName.matches("^\\d+\\.\\d+\\.\\d+$")) {
            LogHelper.warning("Not semantic versioning");
            return null;
        }
        final String[] versionNames = versionName.split("\\.");
        return new Integer[] {
                Integer.valueOf(versionNames[0]),
                Integer.valueOf(versionNames[1]),
                Integer.valueOf(versionNames[2])
        };
    }

    @Nullable
    public static String versionName() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionName;
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static int versionCode() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return -1;
            }

            return packageManager.getPackageInfo(packageName(), PackageManager.GET_META_DATA).versionCode;
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return -1;
        }
    }

    @Nullable
    public static Bitmap icon() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }
            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packageName(), 0);
            if (applicationInfo == null) {
                LogHelper.warning("ApplicationInfo was NULL");
                return null;
            }
            final Drawable drawable = applicationInfo.loadIcon(packageManager);
            return BitmapHelper.fromDrawable(drawable);
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    @Nullable
    public static List<String> permissions() {
        try {
            final PackageManager packageManager = packageManager();
            if (packageManager == null) {
                LogHelper.warning("PackageManager was NULL");
                return null;
            }

            final PackageInfo packageInfo = packageManager.getPackageInfo(packageName(), PackageManager.GET_PERMISSIONS);

            final List<String> list = new ArrayList<>();
            Collections.addAll(list, packageInfo.requestedPermissions);
            return list;
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    @SuppressLint("PackageManagerGetSignatures")
    @Nullable
    public static String signature() {
        final PackageManager packageManager = packageManager();
        if (packageManager == null) {
            LogHelper.warning("PackageManager was NULL");
            return null;
        }

        try {
            final PackageInfo packageInfo = packageManager.getPackageInfo(packageName(), PackageManager.GET_SIGNATURES);
            if (packageInfo == null) {
                LogHelper.warning("PackageInfo was NULL");
                return null;
            }

            final Signature[] signatures = packageInfo.signatures;
            if (signatures == null) {
                LogHelper.debug("No signatures");
                return null;
            }

            return signatures[0].toCharsString();
        }
        catch (final PackageManager.NameNotFoundException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String installer() {
        return installer(packageName());
    }

    @SkHide
    public static String installer(@NonNull final String packageName) {
        return packageManager().getInstallerPackageName(packageName);
    }

    public static boolean installedFromGooglePlay() {
        return installer().equalsIgnoreCase(INSTALLER);
    }

}

File: library/src/main/java/me/shkschneider/skeleton/helper/ApplicationHelper.kt
package me.shkschneider.skeleton.helper

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.Bitmap
import android.os.Build
import me.shkschneider.skeleton.SkeletonApplication
import me.shkschneider.skeleton.ui.BitmapHelper

object ApplicationHelper {

    // ApplicationInfo.loadDefaultIcon(packageManager)
    const val DEFAULT_ICON = android.R.drawable.sym_def_app_icon
    const val INSTALLER = "com.android.vending"

    fun debuggable(): Boolean {
        // <https://stackoverflow.com/a/25517680/603270>
        return SkeletonApplication.DEBUGGABLE
    }

    fun resources(): Resources {
        // Resources.getSystem()
        return ContextHelper.applicationContext().resources
    }

    fun packageName(): String {
        return ContextHelper.applicationContext().packageName
    }

    fun packageManager(): PackageManager {
        return ContextHelper.applicationContext().packageManager
    }

    fun applicationInfo(packageName: String = ApplicationHelper.packageName()): ApplicationInfo? {
        try {
            val packageManager = packageManager()
            return packageManager.getApplicationInfo(packageName, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    fun name(packageName: String = ApplicationHelper.packageName()): String? {
        val label = applicationInfo(packageName)?.loadLabel(packageManager()) ?: run {
            Logger.warning("Label was NULL")
            return null
        }
        return label.toString()
    }

    fun exists(packageName: String = ApplicationHelper.packageName()): Boolean {
        try {
            return (packageManager().getApplicationInfo(packageName, 0) != null)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return false
        }
    }

    fun packageInfo(packageName: String = ApplicationHelper.packageName(), flags: Int = 0): PackageInfo? {
        try {
            val packageManager = packageManager()
            return packageManager.getPackageInfo(packageName, flags)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    fun versionName(packageName: String = ApplicationHelper.packageName()): String? {
        return packageInfo(packageName, PackageManager.GET_META_DATA)?.versionName ?: run {
            Logger.warning("VersionName was NULL")
            return null
        }
    }

    @SuppressLint("NewApi")
    fun versionCode(packageName: String = ApplicationHelper.packageName()): Long? {
        val packageInfo = packageInfo(packageName, PackageManager.GET_META_DATA)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            return packageInfo?.longVersionCode ?: run {
                Logger.warning("LongVersionCode was NULL")
                return null
            }
        } else {
            @Suppress("DEPRECATION")
            return packageInfo?.versionCode?.toLong() ?: run {
                Logger.warning("VersionCode was NULL")
                return null
            }
        }
    }

    fun buildTime(packageName: String = ApplicationHelper.packageName()): Long? {
        return packageInfo(packageName, PackageManager.GET_META_DATA)?.lastUpdateTime ?: run {
            Logger.warning("LastUpdateTime was NULL")
            return null
        }
    }

    fun permissions(packageName: String = ApplicationHelper.packageName()): List<String>? {
        return packageInfo(packageName, PackageManager.GET_PERMISSIONS)?.requestedPermissions?.toList()
                ?: run {
                    Logger.warning("RequestedPermissions was NULL")
                    return null
                }
    }

    @Deprecated("You should use @DrawableRes when possible.")
    fun icon(packageName: String = ApplicationHelper.packageName()): Bitmap? {
        try {
            val packageManager = packageManager()
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            val drawable = applicationInfo.loadIcon(packageManager)
            return BitmapHelper.fromDrawable(drawable)
        } catch (e: PackageManager.NameNotFoundException) {
            Logger.wtf(e)
            return null
        }
    }

    @TargetApi(AndroidHelper.API_19)
    fun clear() {
        Logger.info("clearApplicationUserData()")
        SystemServices.activityManager()?.clearApplicationUserData()
    }

}
